src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -2432,10 +2436,11 @@ namespace QgsWms
   {
     //qgs:%TYPENAME%
     QDomElement typeNameElement = doc.createElement( "qgs:" + typeName /*qgs:%TYPENAME%*/ );
-    typeNameElement.setAttribute( QStringLiteral( "fid" ), typeName + "." + QString::number( feat->id() ) );
+    const QString fid( QgsServerFeatureId::getServerFid( *feat, layer->dataProvider()->pkAttributeIndexes() ) );
+    typeNameElement.setAttribute( QStringLiteral( "fid" ), QStringLiteral( "%1.%2" ).arg( typeName, fid ) );
 
     QgsCoordinateTransform transform;
-    if ( layer && layer->crs() != crs )
+    if ( layer && layer->crs() != crs ) // cppcheck-suppress nullPointerRedundantCheck
```

Comment:
I think the cppcheck is correct:

- either we are sure there is a layer and we can remove the nullptr check here
- or we need to check for a nullptr on line 2439 as well

Reply:
ok

Reply:
Removing the check because it's already checked in client code

Reply:
Can you add a `Q_ASSERT` to make the assumption explicit?

Reply:
nope, the check is correct, there is a case when a nullptr layer is passed.

Reply:
I've added the missing check for the newly introduced fid from data provider, we should be safe now.